src/query/api/v1/handler/prom/prom.go

```diff
@@ -41,26 +41,35 @@ func init() {
 // Options defines options for PromQL handler.
 type Options struct {
 	PromQLEngine *promql.Engine
+	instant bool
+}
```

Comment:
I'm not sure about this mixed approach where some of the struct fields are public, while others are not. A more common approach I've seen was to have private struct with all private fields and public interface for it, e.g.: 
https://github.com/m3db/m3/blob/166bdb36cdb589caa67c3ef2df0b13c02b1d80de/src/dbnode/ratelimit/options.go#L35-L48

Alternatively, instant could be passed as a method argument. It's not very option-y, when you need to construct a handlers with both `instant=true` and `instant=false`.

Reply:
We should be migrating towards functional options: https://github.com/uber-go/guide/blob/master/style.md#functional-options

Reply:
A bit contrived example with error handling could look like:
```
func newDefaultOptions() options {
	return options{
		engine:  mydefaultengine(),
		instant: false,
		foobar:  42,
	}
}

type Option interface {
	apply(*options) error
}

type instantEngineOption struct {
	engine interface{}
}

func (o instantEngineOption) apply(opts *options) error {
	if o.engine == nil {
		return errors.New("invalid engine")
	}
	opts.instant = true
	opts.engine = o.engine
	return nil
}

func WithInstantEngine(algo placement.Algorithm) Option {
	return instantEngineOption{engine: algo}
}

func NewReadHandler(hOpts options.HandlerOptions /*mandatory param*/, opts ...Option) (http.Handler, error) {
	options := newDefaultOptions()
	for _, o := range options {
		if err := o.apply(&options); err != nil {
			return nil, err
		}
	}
	// do something with options
	storageEngine := o.engine
	// ..
}
// ..
h, err := NewReadHandler(hopts, WithInstantEngine(foo), WithFooOption(bar))
```